### ISSUE TITLE
[Improvement] RedisCluster::OPT_SLAVE_FAILOVER distribute readonly commands to the selected node

### ISSUE THREAD

**Issue by Tasiobg** [phpredis/phpredis#1194]
RedisCluster::FAILOVER_DISTRIBUTE is distributing readonly commands between masters and slaves at random. I would like to be able to choose the preferred node in each connection, this way I can choose the closer one to this app server and save some time.

**Comment by yatsukhnenko** [comment]
@Tasiobg, how this 'choosing' should looks like?

**Comment by Tasiobg** [comment]
Thanks for having a look at this suggestion

@yatsukhnenko may be something like this?
```
$obj_cluster = new RedisCluster(null, Array('host:7000', 'host:7001', 'host:7003'));
$obj_cluster->setOption(
    RedisCluster::OPT_SLAVE_FAILOVER, RedisCluster::FAILOVER_PREFERRED_NODE, 'host:7001'
);
```

Some APP are distributed among different availability zones and running behind a load balancer like HAProxy, I would like to be able to choose a preferred node per request because I know which node is closer.

Not sure how good is the example above but this only need to be set once per request and I think setting this parameter within the connection process should be fine.

**Comment by trvrnrth** [comment]
Was there ever any movement on adding a feature like this?

I have a similar setup whereby I would like to direct read requests to a preferred node in order to avoid cross-az charges and latency.

Being able to specify a single node would be sufficient for our current setup, although perhaps an array of preferred nodes would be better for future proofing.

I notice that the session handler allows hosts to be given a weighting. Perhaps something of that ilk would work here?

**Comment by trvrnrth** [comment]
@michael-grunder It looks like you were the last person to touch this area. Do you have any input on whether such a feature is something that would be considered? Or an alternative approach that we might be able to take? I am happy to elaborate on my particular use case if that would help.

I may be able to contribute some time towards adding a feature to accomplish this but would only want to do so if it's design is such that it would be accepted.

**Comment by DemigodCode** [comment]
Would need that too.
Using redis cluster with "remote" nodes has got a poor performance.
Our Website slows down 200ms with cluster configuration.

**Comment by aolley** [comment]
So I've taken a crack at this (see #1896).

Example usage:
```
<?php                                                                                                                                                         

$rc = new RedisCluster(null, ['127.0.0.1:7000']);
$rc->setOption(RedisCluster::OPT_SLAVE_FAILOVER, RedisCluster::FAILOVER_PREFERRED);
$rc->setOption(RedisCluster::OPT_PREFERRED_NODES, ['127.0.0.1:7001', '127.0.0.1:7010']);

// verify preferred nodes are set
var_export($rc->getOption(RedisCluster::OPT_PREFERRED_NODES));
echo "\n";

// set a key (when i ran this, "key1" happened to map the nodes i listed above, allowing me to verify with redis-cli monitor that the preferred nodes were getting the GETs but the other nodes weren't)
$rc->set('key1', 'value');
// get the key
echo $rc->get('key1')."\n";
```

In-so-far as I can make out - this works well, and with only a minor hit to performance (easily outweighed by latency if you're preferring nodes closer to you):

Over gets for 1M keys, performance of fail over modes:
NONE: 28.9us per get
DISTRIBUTE: 35.4us per get
PREFERRED: 36.7us per get

But C isn't my forte - let alone writing php extensions.

**Comment by RafalLukawiecki** [comment]
Does your PR allow for specifying the master node as the preferred one for reads?

I am looking for a very simple scenario of having master and just one replica, each in a different availability zone. I would like the app server in the zone where the master is located to use the master for everything, while the other zone's app server to use the replica for reads (and master for writes).

I realise this can get more complex in case of a failover across the availability zones, but I can solve that by switching CNAMEs when that happens.

**Comment by aolley** [comment]
@RafalLukawiecki yes it does. The preferential selection acts on the full list of nodes, which includes the primary and all its replicas.
